FEATURE: Introduce `Workspace::hasPublishableChanges`

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

Introduces `Workspace::hasPublishableChanges` which will be used for following cases:
- to make `Discard` a noop (instead of looking into the event stream)
- to make `ChangeBaseWorkspace` fail if the workspace has changes (instead of looking into the event stream)
- to make `DeleteWorkspace` fail if the workspace has changes (future pr)
- for node migrations as it should be standalone package and NOT use the change finder ([link](https://github.com/neos/neos-development-collection/blob/97a3588560cd6b0c33c414a70eb7a849b0c226b4/Neos.ContentRepository.NodeMigration/src/NodeMigrationService.php#L200))

Also its good to have some rudimentary changes concept already low level which will hopefully improved at somepoint in Neos X by diffing two content graphs or something.

Note also that the `hasPublishableChanges` will also return true for the case a NodeMigration was run on the workspace and created events like `DimensionShineThroughWasAdded` which are in fact publish able (which is NOT detected via the change projection !!!).

## Implementation:

We decided for a simple `hasChanges` boolean flag on the content stream for now which will be set if there is at least one event that is considered a change: `PublishableToWorkspaceInterface`.
Technically an increment to count the events that are `PublishableToWorkspaceInterface` could also work but as there is no grouping of events per node aggregate id it would be hard to reason about the number and its meaning (it would be completely different to the change projections count)

We leverage the `PublishableToWorkspaceInterface` to ignore events like forking, closing and reopening, as they are not published to another stream. 

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
FEATURE: Introduce `publishableEvents` count on content stream

bwaidelich commented:
> @mhsdesign wrote in slack
> But instead of having a boolean flag i rather have already some simple incrementing count no? That would allow us to get rid of usages of the change projection like when deleting workspaces and saying how many nodes are in the workspace:
`publishableEvents=publishableEvents+1`
To be true a `PublishableToWorkspaceInterface`  does NOT equal a change in the change projection as multiple set properties are deduplicated but i think it would still be a good think to have this low level count available.
The workspace would get a method hasPublishableChanges (wich never true for root workspaces) and maybe something countPublishableChanges though we should probably refrain from the whole changes term here as its already claimed by the change projection.
For example a DimensionShineThroughWasAdded event is also a change in terms for the workspace model but NOT respected by the change projection which only handles nodes.
Maybe we should use hasPublishableEvents and countPublishableEvents to be fully distinct?

One note regarding count vs flag:
A counter would only replace the change projection if we kept an individual count per type of edit (create, update, delete) per document node which would require a separate table.
For the time being I would go with the simple boolean flag and expose only hasPublishableChanges()  on the workspace (since that can always be derived from the separated changes if we track them at some point).
Or do we need Workspace::countPublishableChanges() for anything else?

mhsdesign committed:
TASK: Prefer simple dirty flag over counting publishable changes

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Projection/Feature/ContentStream.php:
```diff
@@ -49,12 +54,18 @@ private function removeContentStream(ContentStreamId $contentStreamId): void
+    private function updateContentStreamVersion(EventInterface&EmbedsContentStreamId $event, Version $version): void
     {
-        $this->dbal->update($this->tableNames->contentStream(), [
+        // todo make fork content stream `EmbedsContentStreamId` but then just ignore it here because we set the version already
```
+1

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -179,6 +179,11 @@ private function handlePublishWorkspace(
         $baseWorkspace = $this->requireBaseWorkspace($workspace, $commandHandlingDependencies);
+        if (!$workspace->hasPublishableChanges()) {
+            // no-op
+            return;
```
I would suggest to throw a dedicated exception in this case to make it really explicit.
We can always decide to catch it in the application (i.e. Neos) if we need to but just ignoring that case is "dangerous" IMO

bwaidelich replied:
@mhsdesign promised to tackle this in a follow-up :)

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -358,14 +353,7 @@ private function handleRebaseWorkspace(
-        );
-
-        if ($rebaseableCommands->isEmpty()) {
+        if (!$workspace->hasPublishableChanges()) {
```
Same here, IMO we should make this case explicit, i.e. throw an exception if there is nothing to rebase

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -438,16 +433,17 @@ private function handlePublishIndividualNodesFromWorkspace(
-        if ($command->nodesToPublish->isEmpty()) {
+        $workspace = $this->requireWorkspace($command->workspaceName, $commandHandlingDependencies);
+        $baseWorkspace = $this->requireBaseWorkspace($workspace, $commandHandlingDependencies);
+        if ($command->nodesToPublish->isEmpty() || !$workspace->hasPublishableChanges()) {
```
and here

mhsdesign committed:
TASK: Do not try catch `countPendingWorkspaceChanges`

... because it would lead to hard to debug code and is not expected to fail

mhsdesign committed:
TASK: Make `ContentStreamWasForked` implement `EmbedsContentStreamId`

... and dont `updateContentStreamVersion` for it

https://neos-project.slack.com/archives/C04PYL8H3/p1730292774760689

bwaidelich reviewed (approved):
This looks great, thanks!
Except for the implicit no-ops for invalid publish/discards, but you wanted to change that in a follow-up IIRC

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/Workspaces/PruneContentStreams.feature:
```diff
@@ -118,28 +122,131 @@ Feature: If content streams are not in use anymore by the workspace, they can be
+      | name   | base | status       | content stream  | publishable changes |
+      | "live" | null | "UP_TO_DATE" | "cs-identifier" | false               |
+
+    When I am in workspace "user-test" and dimension space point {}
+    # FIXME maybe getContentGraph should already throw an exception if the content stream does not exist?
+    Then I expect node aggregate identifier "root-node" to lead to no node
```
todo its a little dirty that `getContentGraph` still works in the detached state (as we just fetch the `currentContentStreamId` but dont validate that it exists.

its definitely interesting to think about this as catchup hooks will need to deal with this in case youd want to listen to all `CreateWorkspace` events and get some information at that time from the content graph or the workspace finder ...

mhsdesign commented:
Removed the unrelated rabbit hole changes into a new pr: https://github.com/neos/neos-development-collection/pull/5333

mhsdesign committed:
TASK: rename `dirty` to `hasChanges`

mhsdesign commented:
we discussed in the weekly today that we want name the column `hasChanges` instead and merge it :)

kitsunet commented:
interesting test failure because the test doesn#t say anytghing about no changes....

kitsunet commented:
you know what, not following my own advice here. This has not too much to do with this change, I will revert my changes and merge this. WE can throw later... (as these changes are now gone here, I added throws for the no ops)

mhsdesign commented:
jup thanks for attempting the change though... ill continue it in https://github.com/neos/neos-development-collection/pull/5337 for beta16 :)
